### PR TITLE
docs: first draft of a blog post for the cdf (#23)

### PR DESCRIPTION
**Pull request by ninjamast3r**
Looking for clarification on direction. Will tailor to that once it's more nailed down.

### CODE REVIEW

**Comment by ninjamast3r** [comment]
Looking for feedback here. Going to have Smitty throw this at the marketing guys and lawyers too.

File: docs/explanations/cdf-blog.md [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+  "success": true,
+  "event_receiver_id": "01HDS785T0V8KTSTDM9XGT33QQ"
+}
+```
+
+### Event Receiver Groups
+
+Event receiver groups can be thought of as gates that control whether an
+artifact advances through the pipeline. Each group comprises multiple receivers.
+Like receivers, groups can cause the generation of Redpanda events. However,
```
Comment (caproven) [inline docs/explanations/cdf-blog.md:139]: See prior comment about event vs message

File: docs/explanations/cdf-blog.md [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+- Show messages on the bus
+
+## Running EPR in Production
+
+Now that you understand the basics, here's a real world example. We'll be starting with a build of the fabulous `my-app` application. At the end of the build process, the build automation will post a passing receipt to EPR. The build automation generates an NVRPP which will be used by the first and subsequent events for this artifact. Downstream watchers consume the successful build event on the Redpanda topic, triggering a security scan, integration tests, and artifact signing. Watchers for each of those tasks invoke them with the NVRPP used in the build event. These three tasks post more events their corresponding receivers that are all contained inside a release group. Once those three receivers have passing events for the NVRPP, the group triggers a release event, that a downstream watcher catches to release the software.
```
Comment (caproven) [inline docs/explanations/cdf-blog.md:168]: Watchers header & downward has paragraphs on same line so harder to give inline feedback.

* "post a passing receipt to EPR" -> "post a passing event to EPR"
* "post more events their corresponding receivers" -> "post more events to their corresponding receivers"

File: docs/explanations/cdf-blog.md [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+- Provide a diagram
+
+## Pitfalls
+
+EPR was and is a highly successful project internally. That doesn't mean we didn't have problems. These are a few issues we ran in to that you can hopefully avoid.
+
+### Lack of Access Control for Receivers and Groups
+
+The first big problem we ran into was that there were no restrictions on who could use receivers and groups. This meant that anyone could post passing events to any receiver, which in turn could trigger any associated groups. Add some lazy event matching, and suddenly you find yourself releasing thousands of artifacts without intending to (yes, this actually happened). We discussed adding serious Role Based Access Control (RBAC) to receivers and groups but ultimately decided not to in favor of development speed. The tradeoff was some unfortunate hacks that persist to this day. Now that EPR is open sourced, we intend implement a more robust solution in the near future.
```
Comment (caproven) [inline docs/explanations/cdf-blog.md:178]: "we intend implement" -> "we intend to implement"

**Review by caproven — COMMENTED** [review]
Couple scattered nits but this looks excellent, great job!

Only notable input is on how NVRPP is introduced. The section for it mentions events well before events are even defined, which may cause jumping around or confusion. If possible, you may want to remove/reduce mentions of events in that section. This may also helps alleviate some redundancy between the NVRPP & Events sections.

File: docs/explanations/cdf-blog.md (hunk 2 above)
Comment (ninjamast3r) [reply docs/explanations/cdf-blog.md]: Hah, I knew there was somewhere I said receipt in here.

**Review by caproven — COMMENTED** [review]
Couple nits, otherwise looking good. I think the event/message phrasing is a lot clearer now so thanks for updating that.

File: docs/explanations/cdf-blog.md [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+or three large ship events per year. Updates were slow and painful. To reverse
+that trend, we needed a way to shorten the development cycle and deliver
+artifacts more quickly.
+The challenge was twofold. First, create a system that could allow disparate
+pieces of our pipeline to communicate and chain together. Second, help R&D shift
+gears from the old software development model to CI/CD.
+
+In a happy, imaginary world somewhere, we could have chained together Github
+actions or equivalent into a working pipeline. No need to write EPR at all.
+Reality was not so kind. Our source code was (and still is) scattered accross
```
Comment (xbcsmith) [inline docs/explanations/cdf-blog.md:22]: accross --> across

File: docs/explanations/cdf-blog.md [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+that was simple enough to work just about anywhere.
+
+Our solution was Event Provenance Registry, or rather, the precursor to it. We
+compared it to duct taping a Raspberry Pi to a rusty tractor. EPR is the glue
+that enabled the rest of the pipeline to really take off.
+
+## How it works
+
+EPR is fairly simple in its operation, but requires some explanation. At a high
+level, EPR collects events based on tasks done by the build pipeline and sends
```
Comment (xbcsmith) [inline docs/explanations/cdf-blog.md:38]: build pipeline --> pipeline

Drop the build

File: docs/explanations/cdf-blog.md [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+}
+```
+
+### Watchers
+
+Watchers are applications (typically microservices) that watch the EPR Redpanda
+topic for messages and then take some action. They use some matching logic,
+defined in EPR's SDK to determine which messages to process. You can match on
+NVRPP, `type`, and `success`. We've written about a dozen or so watchers
+interally that do a variety of things. One of our more popular watchers fires
```
Comment (xbcsmith) [inline docs/explanations/cdf-blog.md:166]: interally --> internally

File: docs/explanations/cdf-blog.md [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+
+## Pitfalls
+
+EPR was and is a highly successful project internally. That doesn't mean we
+didn't have problems. These are a few issues we ran in to that you can hopefully
+avoid.
+
+### Lack of Access Control for Receivers and Groups
+
+The first big problem we ran into was that there were no restrictions on who
```
Comment (xbcsmith) [inline docs/explanations/cdf-blog.md:199]: was that there were no -- > was the lack of

File: docs/explanations/cdf-blog.md [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+to any receiver, which in turn could trigger any associated groups. Add some
+lazy message matching, and suddenly you find yourself releasing thousands of
+artifacts without intending to (yes, this actually happened). We discussed
+adding serious Role Based Access Control (RBAC) to receivers and groups but
+ultimately decided not to in favor of development speed. The tradeoff was some
+unfortunate hacks that persist to this day. Now that EPR is open sourced, we
+intend to implement a more robust solution in the near future.
+
+### Adoption was Difficult
+
```
Comment (xbcsmith) [inline docs/explanations/cdf-blog.md:210]: Probably going to reword this section to soften it up. We will see.

**Review by xbcsmith — CHANGES_REQUESTED** [review]

File: docs/explanations/cdf-blog.md [hunk 9]
```diff
@@ -0,0 +1,284 @@
+# EPR
```
Comment (xbcsmith) [inline docs/explanations/cdf-blog.md:1]: EPR --> Event Provenance Registry (EPR)

**Review by xbcsmith — CHANGES_REQUESTED** [review]

**Review by xbcsmith — APPROVED** [review]

**Comment by ninjamast3r** [comment]
Updated the content with the approved version and fixed the formatting. Ready for a real review. I may do the images in a separate PR.

**Review by xbcsmith — APPROVED** [review]

**Pull request merged**